### PR TITLE
Implements maximum minimum

### PR DESCRIPTION
**Pull request by kaushikcfd**

### CODE REVIEW

File: pytato/array.py [hunk 1]
```diff
@@ -1804,4 +1804,38 @@ def ones(namespace: Namespace, shape: ConvertibleToShape, dtype: Any = float,
...
+            x1.namespace,
+            prim.If(condition, expr1, expr2),
+            bindings={"in0": x1, "in1": x2},
+            shape=x1.shape,
+            dtype=dtype,
+            )
+
+
+def maximum(x1: Array, x2: Array) -> IndexLambda:
+    return _select(x1, x2, ">")
```
Comment (kaushikcfd) [inline pytato/array.py:1832]: Question: Is the `c99.max` more apt here? I think the only problem here might be `NaNs`.

Comment (kaushikcfd) [reply pytato/array.py:1832]: Even `c99`'s fmax does not conform to Numpy's spec when exactly one of the element is a NaN.

File: pytato/array.py [hunk 2]
```diff
@@ -1804,4 +1804,38 @@ def ones(namespace: Namespace, shape: ConvertibleToShape, dtype: Any = float,
 
 # }}}
 
+
+# {{{ (max|min)inimum
+
+def _select(x1: Array, x2: Array, op: str) -> IndexLambda:
```
Comment (inducer) [inline pytato/array.py:1810]: This is pretty close to (the beginning of) an implementation of (the three-argument form of) `np.where`. Should we make it that? (It doesn't have to be complete, but cover shared subset of what you need and what numpy has. Also, it should check that the functionality being used stays within that subset.)

File: pytato/array.py (hunk 1 above)
Comment (inducer) [reply pytato/array.py:1832]: I know it's annoying to code, but personally, I think matching numpy's NaN behavior makes a great deal of sense:
```
>>> import numpy as np
>>> np.maximum(np.nan, 3)
nan
>>> np.maximum(3, np.nan)
nan
>>> np.minimum(3, np.nan)
nan
>>> np.minimum(np.nan, 3)
nan
```

One main use case of this is for norms, and handling it at the pytato level removes a ton of annoyance from the higher ones. Maybe we could push this into loopy?

From `man 3 fmax`:
> If one argument is a NaN, the other argument is returned.

:facepalm:

File: test/test_codegen.py [hunk 3]
```diff
@@ -529,6 +529,27 @@ def test_passsing_bound_arguments_raises(ctx_factory):
...
+
+    namespace = pt.Namespace()
+    x1 = pt.make_data_wrapper(namespace, x1_in)
+    x2 = pt.make_data_wrapper(namespace, x2_in)
+    pt_func = getattr(pt, which)
+    np_func = getattr(np, which)
+
+    _, (y,) = pt.generate_loopy(pt_func(x1, x2),
+            target=pt.LoopyPyOpenCLTarget(queue))()
+    np.testing.assert_allclose(y, np_func(x1_in, x2_in), rtol=1e-6)
```
Comment (inducer) [inline test/test_codegen.py:550]: Add NaN tests? :)

**Review by inducer — COMMENTED** [review]
Thanks for working on this! I've got some comments below that are sure to delight. :duck: :)

**Comment by inducer** [comment]
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

File: pytato/array.py [hunk 4]
```diff
@@ -1804,4 +1804,38 @@ def ones(namespace: Namespace, shape: ConvertibleToShape, dtype: Any = float,
...
+    if x1.shape != x2.shape:
+        raise NotImplementedError("broadcasting not supported")
```
Comment (kaushikcfd) [inline pytato/array.py:1812]: Note to self: Now that broadcasting's implemented use it.

Comment (kaushikcfd) [reply pytato/array.py]: Done. b4abad7

File: test/test_codegen.py (hunk 3 above)
Comment (kaushikcfd) [reply test/test_codegen.py:617]: Done. da95da0

File: pytato/array.py (hunk 2 above)
Comment (kaushikcfd) [reply pytato/array.py]: Yep, pt.where is now  a thing!

File: pytato/array.py (hunk 1 above)
Comment (kaushikcfd) [reply pytato/array.py]: > If one argument is a NaN, the other argument is returned.

Controversial choice of C standards. :roll_eyes:

**Comment by kaushikcfd** [comment]
Tests passing depends on: https://github.com/inducer/loopy/pull/254

File: pytato/array.py [hunk 5]
```diff
@@ -1804,4 +1813,174 @@ def ones(namespace: Namespace, shape: ConvertibleToShape, dtype: Any = float,
...
+
+
+def minimum(x1: Union[Array, Number],
+            x2: Union[Array, Number]) -> Union[Array, Number]:
+    """
+    Returns the elementwise minimum of *x1*, *x2*. *x1*, *x2* being
+    array-like objects that could be broadcasted together. NaNs are propagated.
+    """
+    # https://github.com/python/mypy/issues/3186
+    return where(not_equal(x1, x1), np.NaN,  # type: ignore
```
Comment (kaushikcfd) [inline pytato/array.py:1979]: Incorrect, should implement and use `pt.isnan`. Some compiler flags like -ffast_math can optimize this away.

Comment (kaushikcfd) [reply pytato/array.py]: Done. fe68975

**Comment by inducer** [comment]
https://github.com/inducer/loopy/pull/254 is in, and I've rerun the tests... there still seem to be some issues. Unsubscribing, ping me when ready.

**Comment by kaushikcfd** [comment]
@inducer: All the tests pass now and can be reviewed. (thanks in advance!)

File: pytato/utils.py [hunk 6]
```diff
@@ -87,3 +89,56 @@ def with_indices_for_broadcasted_shape(val: prim.Variable, shape: ShapeType,
...
+        if isinstance(expr, Array):
+            dtypes.append(expr.dtype)
+        else:
+            assert isinstance(expr, Number)
+            dtypes.append(expr)
+
+    return dtypes
+
+
+def broadcasted_binary_op(a1: Union[Array, Number], a2: Union[Array, Number],
```
Comment (inducer) [inline pytato/utils.py:106]:
```suggestion
def broadcast_binary_op(a1: Union[Array, Number], a2: Union[Array, Number],
```
(Merriam Webster says "broadcasted" is also OK, but I'm more familiar with "broadcast".)

File: pytato/utils.py [hunk 7]
```diff
@@ -87,3 +89,56 @@ def with_indices_for_broadcasted_shape(val: prim.Variable, shape: ShapeType,
         return val
     else:
         return val[get_indexing_expression(shape, result_shape)]
+
+
+def extract_dtypes(exprs: List[Union[Array, Number]]) -> List[DtypeOrScalar]:
```
Comment (inducer) [inline pytato/utils.py:94]: Is there a reason this should operate on a list? (vs on just one?)

Comment (inducer) [inline pytato/utils.py:94]: Prefer sequence (if needed) or iterable (if possible) over `List`.

Comment (inducer) [inline pytato/utils.py:94]: `extract_dtype` is a weird name for something that returns scalars unmodified.

File: pytato/array.py [hunk 8]
```diff
@@ -1804,4 +1800,189 @@ def ones(namespace: Namespace, shape: ConvertibleToShape, dtype: Any = float,
...
+                       dtype=np.bool8,
+                       bindings={"_in0": x})
+
+# }}}
+
+
+# {{{ where
+
+def where(condition: Union[Array, Number],
+          x: Union[Array, Number], y: Union[Array, Number]) -> Union[Array, Number]:
```
Comment (inducer) [inline pytato/array.py:1914]: `where` actually also has a single-argument form. Maybe raise an intelligible error if that's what the user wanted?

File: pytato/array.py [hunk 9]
```diff
@@ -1804,4 +1800,189 @@ def ones(namespace: Namespace, shape: ConvertibleToShape, dtype: Any = float,
...
+    def _update_bindings_and_get_expr(arr: Union[Array, Number],
+                                      bnd_name: str) -> ScalarExpression:
+
+        if isinstance(arr, Number):
+            return arr
+
+        bindings[bnd_name] = arr
+        return utils.with_indices_for_broadcasted_shape(prim.Variable(bnd_name),
+                                                        arr.shape,
+                                                        result_shape)
```
Comment (inducer) [inline pytato/array.py:1947]: This pattern is repeated in `utils`. Is it worth factoring it out?

**Review by inducer — COMMENTED** [review]
Thanks!

**Comment by inducer** [comment]
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

File: pytato/array.py (hunk 9 above)
Comment (kaushikcfd) [reply pytato/array.py:1958]: Indeed. 501fa85b4dbbdd4ae1654d3943bd49eb75e88405

File: pytato/array.py (hunk 8 above)
Comment (kaushikcfd) [reply pytato/array.py]: Yep, had not accounted for that. Thanks! bf89f130908aa4e07bc50e4e721fa0ccebb31a6f

File: pytato/utils.py (hunk 7 above)
Comment (kaushikcfd) [reply pytato/utils.py]: Agreed. 501fa85b4dbbdd4ae1654d3943bd49eb75e88405

Comment (kaushikcfd) [reply pytato/utils.py]: Done. 501fa85b4dbbdd4ae1654d3943bd49eb75e88405

Comment (kaushikcfd) [reply pytato/utils.py]: I preferred this over writing list comprehensions everywhere.

File: pytato/utils.py (hunk 6 above)
Comment (kaushikcfd) [reply pytato/utils.py]: Sure. 6566d74a9d818b3ea9b423d9c4c8c6fbd2c83c14

File: pytato/array.py [hunk 10]
```diff
@@ -1804,4 +1800,203 @@ def ones(namespace: Namespace, shape: ConvertibleToShape, dtype: Any = float,
...
```
Comment (inducer) [inline pytato/array.py:1958]:
```suggestion
```

**Comment by inducer** [comment]
LGTM. Thanks!

File: pytato/array.py [hunk 11]
```diff
@@ -1804,4 +1800,192 @@ def ones(namespace: Namespace, shape: ConvertibleToShape, dtype: Any = float,
...
+
+    x_dtype = x.dtype if isinstance(x, Array) else np.dtype(type(x))
+    y_dtype = y.dtype if isinstance(y, Array) else np.dtype(type(y))
+    dtype = np.find_common_type([x_dtype, y_dtype], [])
+
+    # }}}
+
+    result_shape = utils.get_shape_after_broadcasting([condition, x, y])
+
+    bindings = {}
```
Comment (kaushikcfd) [inline pytato/array.py:1947]:
```suggestion
    bindings: Dict[str, Array] = {}
```

**Pull request merged**